jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/naming/KeyUtils.java

```diff
@@ -215,7 +210,15 @@ public static String getConcatedTypeNameValues(List<String> typeNames, String ty
 	 */
 	public static String getConcatedTypeNameValues(Query query, List<String> typeNames, String typeName) {
 		Set<String> queryTypeNames = query.getTypeNames();
-		if (queryTypeNames != null && queryTypeNames.size() > 0) {
+		if (query.isUseAllTypeNames()) {
```

Comment:
Yep, that's the idea. I do not have the details all that clear in my head (only one of my 2 neurons is still working properly). The final goal would be to have the correct implementation injected at runtime in the consumer (probably the `Query` class which should expose less of its internal state and provide higher level abstractions ...).